Wire Domain and Reputation Filtering

rdig opened the pull request:
## Description

This PR wires in, mostly in `ColonyHome` domain filtering, as well as member by reputation filtering _(which are related)_.

Alongside this, it will properly set up the members by domains route _(once #2295 lands)_ with the proper members filtered.

**Changes** 
- [x] `ColonyHome` set up domain filtering by either location query or state
- [x] Added `colonyMembersWithReputation` client resolver
- [x] Fix `UserPermissions` to only show permissions if available
- [x] Various fixes to the `Members` page
- [x] `Members` page now can be accessed based on a domain id
- [x] Some minor fixes that I've found along the way

**Demo**

![cs-filter-domain-reputation](https://user-images.githubusercontent.com/1193222/100486538-eb2e9800-310c-11eb-9a84-3b68125d9782.gif)

### Testing Notes

Note, that while testing this locally you will run into an edge case where the number of users displayed in a particular domain is bigger the the "total" number of subscribed users. That is because we're using a [mock reputation oracle](https://github.com/JoinColony/mock-oracle).

I've described this issue in more detail on Discord:
- https://discord.com/channels/562263648173555742/611581703026245689/781904708053893130
- https://discord.com/channels/562263648173555742/611581703026245689/781904759372120074

Resolves DEV-55

rdig commented:
I have no idea why the tests suddenly decided to fail. It's not something introduced here. Not even the underlying packages were updated. Honestly, this is just bizarre.

![Screenshot from 2020-11-28 00-39-53](https://user-images.githubusercontent.com/1193222/100487667-87a76900-3112-11eb-8cbe-c6f177a6ca43.png)

So the `node-proper-lockfile` library that `jest` uses under the hood fails. 

Now the owners of `node-proper-lockfile` claim the issue lies with `jest` as they didn't implement the spec correctly, and because of this, refuse to fix it: https://github.com/moxystudio/node-proper-lockfile/issues/90

There's even a PR to fix this, but apparently, they don't want to merge it because of the above reason _(which, frankly, is understandable)_: https://github.com/moxystudio/node-proper-lockfile/pull/91

I'll have a look on Monday to see if I can't get to the bottom of this, or, at the very least, find a temporary solution for this. _(well... I know of one, but I want a "proper" one)_

rdig commented:
[This](https://github.com/JoinColony/colonyDapp/pull/2302#issuecomment-735005308) gets even weirder: https://github.com/JoinColony/colonyDapp/pull/2302#issuecomment-735005308 since this is not reproducible on the base branch.

So that makes me want to reach the conclusion that it's something I introduced?

But I didn't touch anything related to either those tests or the components those tests, well... test.

Truly weird

alicjakujawa commented on src/data/resolvers/colony.ts:
```diff
@@ -49,6 +80,45 @@ export const colonyResolvers = ({
+       * Let say that then, a user aquires some reputation, which, after a time, decays back to zero.
+       * That user will be returned in the below query.
+       *
+       * Maybe at some point we want to consider filering them out.
```
filtering #typo 😈

alicjakujawa commented on src/modules/core/components/MembersList/MembersListItem.tsx:
```diff
@@ -110,21 +112,27 @@ const MembersListItem = <U extends AnyUser = AnyUser>(props: Props<U>) => {
-          {displayName && (
-            <span className={styles.displayName} title={displayName}>
-              {displayName}
+          {userProfile.profile.displayName && (
```
Are we totally sure that userProfile won't be null?

rdig replied:
I think so...? But I'll check anyway :)

rdig replied:
Apparently, I was correct. `useUser` will always return a correct user object, with **at least** the wallet address populated.

Eg:
```js
{
  profile: {
    walletAddress: "0x0....."
  }
}
```

So that means we can always expect that `userProfile` is set, and we can always destructure from it, without either TS screaming at us or the React component throwing an error.

Here's two `userProfile`'s logged, one has a username registered, the otherone doesn't:
![Screenshot from 2020-11-29 13-43-56](https://user-images.githubusercontent.com/1193222/100540954-5f6e5600-3249-11eb-9943-76081a0970b4.png)

alicjakujawa commented on src/modules/core/components/MembersList/MembersListItem.tsx:
```diff
@@ -110,21 +112,27 @@ const MembersListItem = <U extends AnyUser = AnyUser>(props: Props<U>) => {
-              <UserMention hasLink={false} username={username} />
+              <UserMention
+                hasLink={false}
+                username={userProfile.profile.username}
```
i feel that it would be better to destruct this object once at the top
{username, displayName} = userProfile.profile

rdig replied:
Sure

alicjakujawa commented:
@rdig So the thing which makes node build filing is the change with exporting/importing useTokenInfo
If you revert it old state - so importing like import { useTokenInfo } from '~utils/hooks/useTokenInfo';
Tests are green again

rdig commented:
> @rdig So the thing which makes node build failing is the change with exporting/importing useTokenInfo
If you revert it to old state - so importing like that: import { useTokenInfo } from '~utils/hooks/useTokenInfo';
Tests are green again

Oh. Wow. Nice going finding this out! I owe you one :tumbler_glass:

alicjakujawa commented on src/modules/dashboard/components/ColonyHome/ColonyHome.tsx:
```diff
@@ -83,11 +83,16 @@ const ColonyHome = ({ match, location }: Props) => {
+  );
+
+  const filteredDomainId = domainIdFilter
+    ? domainIdFilter || COLONY_TOTAL_BALANCE_DOMAIN_ID
```
```suggestion
   const filteredDomainId = domainIdFilter || COLONY_TOTAL_BALANCE_DOMAIN_ID
```

rdig replied:
Nice catch. 

This had other logic in it initially, but apparently, I failed to clean it up properly.

alicjakujawa commented on src/modules/dashboard/components/ColonyHome/ColonyMembers/ColonyMembers.tsx:
```diff
@@ -22,64 +24,68 @@ const MSG = defineMessages({
+    }
+    return `${baseRoute}/${currentDomainId}`;
+  }, [currentDomainId, colonyName]);
+
```
i feel that we don't need useMemo here, its quite inexpensive operation, we're just returning correct string

rdig replied:
Ok

rdig commented:
@alicjakujawa this is ready for another look:
- fixed the things you pointed out in the review
- fixed the `tokenInfo` hook imports/exports that also fixed the tests. Thanks for this again!

alicjakujawa commented:
Code looks great. I have only w few doubts about functionality itself. Tested locally a few scenario and im not sure if this working as expected. 
1. Going from Colony Home by clicking on Members redirect me to Members page with All domains selected. This is **CORRECT**:
<img width="1084" alt="Screenshot 2020-11-30 at 07 12 33" src="https://user-images.githubusercontent.com/13069555/100574969-71003e00-32db-11eb-8c76-1515a2e4f16b.png">
2. Selecting "root" from dropdown list is showing above item + random users (i believe its because of our oracle mock server) So im guessing its still **CORRECT**
<img width="1003" alt="Screenshot 2020-11-30 at 07 14 24" src="https://user-images.githubusercontent.com/13069555/100575100-b1f85280-32db-11eb-8934-6614f1c19453.png">
3. Writing random number in url ex. "/colony/a/members/3" will display all mock-oracle data, and keep dropdown empty, this seems to be a bug:
<img width="1034" alt="Screenshot 2020-11-30 at 07 15 46" src="https://user-images.githubusercontent.com/13069555/100575253-fdaafc00-32db-11eb-9e61-e0c8a300a40b.png">
4. When we write letters in url as param ex. "/colony/a/members/ab" - we're fallback to "all domains" so its correct

What i think we're missing is, to check if param is a number, and if SO then check if this domain number actually exits, if not -> fallback to all domains

rdig commented:
> 3. Writing random number in url ex. "/colony/a/members/3" will display all mock-oracle data, and keep dropdown empty, this seems to be a bug:
Yes this is a bug, I'll fix it

> What i think we're missing is, to check if param is a number, and if SO then check if this domain number actually exits, if not -> fallback to all domains
Hmmm. I'm reluctant about this since it's kinda an extra operation just for one edge case. But I guess you're right, it's an edge case and we need to cover it.

rdig commented on src/data/resolvers/colony.ts:
```diff
@@ -21,10 +29,33 @@ import {
+    colonyAddress,
+  );
+  const { skillId } = await colonyClient.getDomain(domainId);
+  /*
+   * @TODO This is coming from colonyJS as I forgot to adjust the proper types when
+   * creating this common extension.
+   * I'll put in a PR to fix this at some point
+   */
+  // @ts-ignore
```
I've put a PR to fix in this colonyJS: https://github.com/JoinColony/colonyJS/pull/480

rdig commented:
@alicjakujawa fixed via 9b964e5

rdig commented:
Refixed via e4c8622

alicjakujawa reviewed (approved):
Perfect! 🔥
